#include <iostream>
#include <memory>
#include <string>
#include <thread>
#include <vector>
#include <atomic>
#include <boost/asio.hpp>
#include <windows.h> // Windows API Header
#include <unordered_map>
#include <mutex>

#include <fstream>      
#include <cstring>      // memset ì‚¬ìš©ì„ ìœ„í•´ ì¶”ê°€
#include <recastnavigation/DetourNavMesh.h>         // í•µì‹¬ êµ¬ì¡°ì²´ ì •ì˜ í¬í•¨
#include <recastnavigation/DetourNavMeshBuilder.h>

#include "protocol.pb.h"
#include "PacketDispatcher.h"

// ê¸°ì¡´ ì‘ì„±í•œ ê²Œì„ ë¡œì§ í—¤ë”ë“¤ì€ ì´ ìë¦¬ ê·¸ëŒ€ë¡œ ìœ ì§€
#include "Zone/Zone.h"
#include "Monster/Monster.h"
#include "Pathfinder/Pathfinder.h"
#include "Pathfinder/MapGenerator.h"
#include "Monster/MonsterManager.h"


#pragma pack(push, 1)
struct PacketHeader {
    uint16_t size;
    uint16_t id;
};
#pragma pack(pop)

using boost::asio::ip::tcp;

// ==========================================
// â˜… ê²Œì„ ì›”ë“œ ìƒíƒœ ë° Zone ê´€ë¦¬
// ==========================================
//Zone g_zone(1000, 1000, 50); // ì˜ˆ: 1000x1000 ë§µ, ê²©ì í¬ê¸° 50
std::unique_ptr<Zone> g_zone;

// â˜… AI ì—°ë™ì„ ìœ„í•œ ì „ì—­ ë³€ìˆ˜
NavMesh g_navMesh;
std::vector<std::shared_ptr<Monster>> g_monsters;

struct PlayerInfo {
    uint64_t uid;
    float x, y;
};

std::unordered_map<std::string, PlayerInfo> g_playerMap;  // account_id -> PlayerInfo
std::unordered_map<uint64_t, std::string> g_uidToAccount; // uid -> account_id
std::mutex g_gameMutex;
uint64_t g_uidCounter = 1; // Zoneì— ë„£ì„ ê³ ìœ  ë²ˆí˜¸ ë°œê¸‰ìš©

// ==========================================
// 1. ì „ì—­ ë³€ìˆ˜ ë° ë””ìŠ¤íŒ¨ì²˜ ì„¤ì •
// ==========================================
class GatewaySession;
PacketDispatcher<GatewaySession> g_s2s_gateway_dispatcher;

// ê²Œì´íŠ¸ì›¨ì´ ì ‘ì† ê°œìˆ˜ ì¹´ìš´í„° (ì¼ë°˜ì ìœ¼ë¡œ 1ê°œì§€ë§Œ í™•ì¥ì„±ì„ ìœ„í•´ ìœ ì§€)
static std::atomic<int> g_connected_gateways{ 0 };


// ==========================================
// 2. GatewaySession: Gatewayë¡œë¶€í„°ì˜ S2S í†µì‹ (ìˆ˜ì‹ /ì†¡ì‹ ) ë‹´ë‹¹
// ==========================================
class GatewaySession : public std::enable_shared_from_this<GatewaySession> {
private:
    tcp::socket socket_;
    PacketHeader header_;
    std::vector<char> payload_buf_;

public:
    GatewaySession(tcp::socket socket) noexcept : socket_(std::move(socket)) {}

    void start() {
        ReadHeader();
    }

    void Send(uint16_t pktId, const google::protobuf::Message& msg) {
        std::string payload;
        msg.SerializeToString(&payload);
        PacketHeader header;
        header.size = static_cast<uint16_t>(sizeof(PacketHeader) + payload.size());
        header.id = pktId;

        auto send_buf = std::make_shared<std::vector<char>>(header.size);
        memcpy(send_buf->data(), &header, sizeof(PacketHeader));
        memcpy(send_buf->data() + sizeof(PacketHeader), payload.data(), payload.size());

        auto self(shared_from_this());
        boost::asio::async_write(socket_, boost::asio::buffer(*send_buf),
            [this, self, send_buf](boost::system::error_code ec, std::size_t) {
                if (ec) std::cerr << "[GameServer] Gatewayë¡œ S2S íŒ¨í‚· ì „ì†¡ ì‹¤íŒ¨\n";
            });
    }

private:
    void ReadHeader() {
        auto self(shared_from_this());
        boost::asio::async_read(socket_, boost::asio::buffer(&header_, sizeof(PacketHeader)),
            [this, self](boost::system::error_code ec, std::size_t length) {
                if (!ec) {
                    if (header_.size < sizeof(PacketHeader) || header_.size > 4096) return;
                    uint16_t payload_size = static_cast<uint16_t>(header_.size - sizeof(PacketHeader));

                    if (payload_size == 0) {
                        // â˜… ì ì¬ì  ì—ëŸ¬ ì‚¬ì „ ì°¨ë‹¨: selfì˜ ë³µì‚¬ë³¸ì„ ë§Œë“¤ì–´ ë„˜ê²¨ì¤ë‹ˆë‹¤.
                        auto session_ptr = self;
                        g_s2s_gateway_dispatcher.Dispatch(session_ptr, header_.id, nullptr, 0);
                        ReadHeader();
                    }
                    else {
                        payload_buf_.resize(payload_size);
                        ReadPayload(payload_size);
                    }
                }
                else {
                    int current_count = --g_connected_gateways;
                    std::cout << "[GameServer] GatewayServerì™€ì˜ S2S ì—°ê²° í•´ì œë¨. (í˜„ì¬ ì—°ê²°ëœ Gateway: " << current_count << "ê°œ)\n";
                }
            });
    }

    void ReadPayload(uint16_t payload_size) {
        auto self(shared_from_this());
        boost::asio::async_read(socket_, boost::asio::buffer(payload_buf_.data(), payload_size),
            [this, self, payload_size](boost::system::error_code ec, std::size_t length) {
                if (!ec) {
                    // â˜… ì ì¬ì  ì—ëŸ¬ ì‚¬ì „ ì°¨ë‹¨: ë³µì‚¬ë³¸ì„ ë§Œë“¤ì–´ ë„˜ê²¨ì¤ë‹ˆë‹¤.
                    auto session_ptr = self;
                    g_s2s_gateway_dispatcher.Dispatch(session_ptr, header_.id, payload_buf_.data(), payload_size);
                    ReadHeader();
                }
            });
    }
};

// ==========================================
// ì´ë™ íŒ¨í‚· í•¸ë“¤ëŸ¬ (Zone ë¡œì§ ê²°í•©)
// ==========================================
void Handle_GatewayGameMoveReq(std::shared_ptr<GatewaySession>& session, char* payload, uint16_t payloadSize) {
    Protocol::GatewayGameMoveReq req;
    if (req.ParseFromArray(payload, payloadSize)) {
        std::string acc_id = req.account_id();
        float new_x = req.x();
        float new_y = req.y();

        std::vector<uint64_t> aoi_uids;

        {
            std::lock_guard<std::mutex> lock(g_gameMutex);

            // 1. ì²˜ìŒ ì´ë™í•˜ëŠ”(ì…ì¥í•œ) ìœ ì €ë¼ë©´ Zoneì— ë“±ë¡
            if (g_playerMap.find(acc_id) == g_playerMap.end()) {
                uint64_t new_uid = g_uidCounter++;
                g_playerMap[acc_id] = { new_uid, new_x, new_y };
                g_uidToAccount[new_uid] = acc_id;

                g_zone->EnterZone(new_uid, new_x, new_y);
                std::cout << "[GameServer] ìœ ì €(" << acc_id << ") ìµœì´ˆ Zone ì§„ì… (UID:" << new_uid << ")\n";
            }
            // 2. ì´ë¯¸ ìˆëŠ” ìœ ì €ë¼ë©´ ì¢Œí‘œ Update
            else {
                auto& info = g_playerMap[acc_id];
                g_zone->UpdatePosition(info.uid, info.x, info.y, new_x, new_y);
                info.x = new_x;
                info.y = new_y;
            }

            // 3. ë‚´ ì£¼ë³€(AOI)ì— ìˆëŠ” ìœ ì € ëª©ë¡ ì¶”ì¶œ
            aoi_uids = g_zone->GetPlayersInAOI(new_x, new_y);
        }

        // 4. Gatewayì—ê²Œ "ì´ ìœ ì €ë“¤ì—ê²Œë§Œ ë¿Œë ¤!" ë¼ê³  íŒ¨í‚· ì „ì†¡
        Protocol::GameGatewayMoveRes s2s_res;
        s2s_res.set_account_id(acc_id);
        s2s_res.set_x(new_x);
        s2s_res.set_y(new_y);
        s2s_res.set_z(req.z());
        s2s_res.set_yaw(req.yaw());

        // [ìˆ˜ì •] ìœ ì €ì™€ ëª¬ìŠ¤í„° ìˆ«ìë¥¼ ë”°ë¡œ ì„¸ê¸° ìœ„í•œ ë³€ìˆ˜
        int user_count = 0;
        int monster_count = 0;

        std::lock_guard<std::mutex> lock(g_gameMutex);
        for (uint64_t target_uid : aoi_uids) {
            // 10000 ë¯¸ë§Œì€ ìœ ì €, ì´ìƒì€ ëª¬ìŠ¤í„°ë¡œ ë¶„ë¥˜í•˜ì—¬ ì¹´ìš´íŒ…
            if (target_uid < 10000) {
                user_count++;

                // ìœ ì €ì¸ ê²½ìš°ì—ë§Œ íŒ¨í‚· ìˆ˜ì‹  ëŒ€ìƒ(target_account_ids)ì— ë„£ìŠµë‹ˆë‹¤.
                auto it = g_uidToAccount.find(target_uid);
                if (it != g_uidToAccount.end()) {
                    s2s_res.add_target_account_ids(it->second);
                }
            }
            else {
                monster_count++;
            }
        }

        // Gatewayë¡œ ì „ë‹¬
        session->Send(Protocol::PKT_GAME_GATEWAY_MOVE_RES, s2s_res);
        // [ìˆ˜ì •] ë­‰ëš±ê·¸ë ¤ì§„ ë¡œê·¸ë¥¼ ìœ ì €ì™€ ëª¬ìŠ¤í„°ë¡œ ë¶„ë¦¬í•´ì„œ ì¶œë ¥í•©ë‹ˆë‹¤.
        std::cout << "[GameServer] ìœ ì €(" << acc_id << ") ì´ë™ ì™„ë£Œ -> AOI ìˆ˜ì‹  ëŒ€ìƒ: ìœ ì € " << user_count << "ëª…, ëª¬ìŠ¤í„° " << monster_count << "ë§ˆë¦¬\n";
    }
}

// ==========================================
// [GameServer] ìœ ì € í‡´ì¥ ì²˜ë¦¬ í•¸ë“¤ëŸ¬
// ==========================================
void Handle_GatewayGameLeaveReq(std::shared_ptr<GatewaySession>& session, char* payload, uint16_t payloadSize) {
    Protocol::GatewayGameLeaveReq req;
    if (req.ParseFromArray(payload, payloadSize)) {
        std::string acc_id = req.account_id();

        std::lock_guard<std::mutex> lock(g_gameMutex);

        // 1. í•´ë‹¹ ìœ ì €ê°€ GameServerì— ì¡´ì¬í•˜ëŠ”ì§€ í™•ì¸
        auto it = g_playerMap.find(acc_id);
        if (it != g_playerMap.end()) {
            uint64_t uid = it->second.uid;
            float last_x = it->second.x;
            float last_y = it->second.y;

            // 2. Zone(ê³µê°„)ì—ì„œ ìœ ì €ì˜ ë¬¼ë¦¬ì  ì‹¤ì²´ ì‚­ì œ
            g_zone->LeaveZone(uid, last_x, last_y);

            // 3. GameServerì˜ ê´€ë¦¬ ë§µì—ì„œ ì™„ì „íˆ ì‚­ì œ
            g_uidToAccount.erase(uid);
            g_playerMap.erase(it);

            std::cout << "[GameServer] ğŸ‘» ìœ ì €(" << acc_id << ", UID:" << uid << ") í‡´ì¥ ì™„ë£Œ. Zoneì—ì„œ ìœ ë ¹ ë°ì´í„° ì‚­ì œë¨.\n";
        }
    }
}

// ==========================================
// 3. GameNetworkServer: 9000ë²ˆ í¬íŠ¸ì—ì„œ Gatewayì˜ ì ‘ì†(Accept) ëŒ€ê¸°
// ==========================================
class GameNetworkServer {
    tcp::acceptor acceptor_;
public:
    GameNetworkServer(boost::asio::io_context& io_context, short port)
        : acceptor_(io_context, tcp::endpoint(tcp::v4(), port)) {
        do_accept();
    }
private:
    void do_accept() {
        acceptor_.async_accept([this](boost::system::error_code ec, tcp::socket socket) {
            if (!ec) {
                int current_count = ++g_connected_gateways;
                std::cout << "[GameServer] S2S í†µì‹ : GatewayServer ì ‘ì† í™•ì¸ ì™„ë£Œ! (í˜„ì¬ ì—°ê²°ëœ Gateway: " << current_count << "ê°œ)\n";
                std::make_shared<GatewaySession>(std::move(socket))->start();
            }
            else {
                std::cerr << "[Error] Gateway Accept ì‹¤íŒ¨: " << ec.message() << "\n";
            }
            do_accept();
            });
    }
};

// ==========================================
// 4. ë©”ì¸ í•¨ìˆ˜: ìŠ¤ë ˆë“œ í’€ êµ¬ì„± ë° ì„œë²„ ì‹¤í–‰
// ==========================================
int main() {
    // ìœˆë„ìš° ì½˜ì†” í•œê¸€ ê¹¨ì§ ë°©ì§€
    SetConsoleOutputCP(CP_UTF8);

    // ì¶”ê°€: í•œê¸€ ì„¸íŒ…ì´ ëë‚œ ì•ˆì „í•œ íƒ€ì´ë°ì— Zoneì„ ìƒì„±í•©ë‹ˆë‹¤!
    g_zone = std::make_unique<Zone>(1000, 1000, 50);

    // [ì¶”ê°€] íŒŒì¼ì´ ì—†ìœ¼ë©´ ì¦‰ì„ì—ì„œ ë§Œë“¤ì–´ì£¼ëŠ” ì œë„ˆë ˆì´í„° ê°€ë™!
    GenerateDummyMapFile("dummy_map.bin");

    // ---------------------------------------------------------
    // ë§µ ë°ì´í„° ë¡œë“œ ë° ëª¬ìŠ¤í„° ìŠ¤í°
    // ---------------------------------------------------------
    g_navMesh.LoadNavMeshFromFile("dummy_map.bin");

    // 3. ëª¬ìŠ¤í„° ìŠ¤í° ë° AI ì‹œìŠ¤í…œ ê°€ë™ (ë¶„ë¦¬ëœ ëª¨ë“ˆ í˜¸ì¶œ)
    InitMonsters();
    StartAITickThread();
    
    // â˜… ë””ìŠ¤íŒ¨ì²˜ ë“±ë¡
    g_s2s_gateway_dispatcher.RegisterHandler(Protocol::PKT_GATEWAY_GAME_MOVE_REQ, Handle_GatewayGameMoveReq);
    g_s2s_gateway_dispatcher.RegisterHandler(Protocol::PKT_GATEWAY_GAME_LEAVE_REQ, Handle_GatewayGameLeaveReq);

    // TODO: ê°œë°œìë‹˜ì˜ ê¸°ì¡´ ì´ˆê¸°í™” ì½”ë“œ (Zone ìƒì„±, Monster ë°°ì¹˜ ë“±)ê°€ ë“¤ì–´ê°ˆ ìë¦¬ì…ë‹ˆë‹¤.
    // InitZone();
    // InitMonsters();

    try {
        boost::asio::io_context io_context;

        // 1. S2S ì„œë²„ ê°ì²´ ìƒì„± (í¬íŠ¸: 9000)
        GameNetworkServer server(io_context, 9000);
        std::cout << "[System] ì½”ì–´ ê²Œì„ ë¡œì§ ì„œë²„ ê°€ë™ (Port: 9000) Created by Jeong Shin Young\n";

        // 2. CPU ì½”ì–´ ê°œìˆ˜ì— ë§ì¶° ìŠ¤ë ˆë“œ ê°œìˆ˜ ì„¤ì •
        unsigned int thread_count = std::thread::hardware_concurrency();
        if (thread_count == 0) thread_count = 4;
        std::cout << "[System] ì›Œì»¤ ìŠ¤ë ˆë“œ ê°œìˆ˜ ì„¤ì •: " << thread_count << "ê°œ\n";

        // 3. ìŠ¤ë ˆë“œ í’€ ìƒì„± ë° io_context.run() ì‹¤í–‰
        std::cout << "[System] ì—¬ëŸ¬ ìŠ¤ë ˆë“œì—ì„œ io_context.run()ì„ í˜¸ì¶œí•˜ì—¬ ìŠ¤ë ˆë“œ í’€ì„ êµ¬ì„±í•©ë‹ˆë‹¤...\n";
        std::vector<std::thread> threads;
        for (unsigned int i = 0; i < thread_count; ++i) {
            threads.emplace_back([&io_context]() {
                io_context.run();
                });
        }
        std::cout << "[System] ìŠ¤ë ˆë“œ í’€ êµ¬ì„± ì™„ë£Œ.\n";

        // 4. ë©”ì¸ ìŠ¤ë ˆë“œ ëŒ€ê¸° (join)
        std::cout << "=================================================\n";
        std::cout << "[System] GatewayServerì˜ S2S ì ‘ì†ì„ ê¸°ë‹¤ë¦¬ëŠ” ì¤‘...\n";

        for (auto& t : threads) {
            if (t.joinable()) {
                t.join();
            }
        }
    }
    catch (std::exception& e) {
        std::cerr << "[Error] ì˜ˆì™¸ ë°œìƒ: " << e.what() << "\n";
    }

    std::cout << "[System] ì„œë²„ê°€ ì•ˆì „í•˜ê²Œ ì¢…ë£Œë˜ì—ˆìŠµë‹ˆë‹¤.\n";
    return 0;
}